src/coreclr/scripts/cse_ml/jitml/jit_cse.py

```diff
@@ -0,0 +1,249 @@
+
+        # update action, ensure we have an up to date observation
+        info['action'] = action
+        del info['observation']
+
+        # Note that we have not yet updated the info dictionary for previous and current, which means
+        # info['current'] is the previous method at this point.  We do not update info's previous/current
+        # until we are sure the method is JIT'ed successfully.
+        previous = info['current']
+
+        # Ensure the selected action is valid.
+        info['action_is_valid'] = self._is_valid_action(action, previous)
+        if info['action_is_valid']:
+            current = self._jit_method_with_cleanup(info['method_index'], JitMetrics=1, JitRLHook=1,
+                                    JitRLHookCSEDecisions=previous.cses_chosen + [action])
+
+            if current is not None:
+                observation = self.get_observation(current)
+                truncated = False
+                terminated = not current.cse_candidates or action is None
+                reward = self.get_rewards(previous, current)
+
+                info['previous'] = previous
+                info['current'] = current
+
+            else:
+                # Don't set current or observation, as we should not be using them.
+                observation = last_info['observation']
+                truncated = True
+                terminated = False
+                reward = INVALID_ACTION_PENALTY
+
+        else:
+            # action was invalid
+            info['invalid_actions'] += 1
+
+            truncated = info['invalid_actions'] >= INVALID_ACTION_LIMIT
+            terminated = False
+            observation = last_info['observation']
+            reward = INVALID_ACTION_PENALTY
+
+        info['observation'] = observation
+        info['total_reward'] += reward
+        info['terminated'] = terminated
+        info['truncated'] = truncated
+
+        # These are reported only once, when the episode is done.
+        if terminated:
+            info['heuristic_score'] = info['heuristic_method'].perf_score
+            info['no_cse_score'] = info['no_cse_method'].perf_score
+            info['total_reward'] = info['total_reward']
+            info['invalid_actions'] = info['invalid_actions']
+            if 'current' in info:
+                info['final_score'] = info['current'].perf_score
+
+        self.last_info = info
+        return observation, reward, terminated, truncated, info
+
+    def get_rewards(self, prev_method : MethodContext, curr_method : MethodContext):
+        """Returns the reward based on the change in performance score."""
+        prev = prev_method.perf_score
+        curr = curr_method.perf_score
+
+        # should not happen
+        if np.isclose(prev, 0.0):
+            return 0.0
+
+        return REWARD_SCALE * (prev - curr) / prev
+
+    def _is_valid_action(self, action, method):
+        # Terminating is only valid if we have performed a CSE.  Doing no CSEs isn't allowed.
```

Comment:
Is this because you track the "no cse" cases separately, so when learning you're always doing some cses?

There will certainly be some instances where doing no cses is the best policy.

Reply:
My overall goal with this checkin is to get something relatively simple and understandable as the baseline for future work.  In this case, my (intentionally) simple reward function isn't capable of understanding an initial "no" choice without adding extra complexity.

A more refined version of this project can and will handle the case where we choose no CSEs to perform, but I did not want to overcomplicate the initial version.